File: app/schemas/calendar_import.py
from datetime import datetime
from typing import Optional
from pydantic import BaseModel


class CalendarFeedCreate(BaseModel):
    url: str


class CalendarFeedResponse(BaseModel):
    id: int
    url: str
    name: Optional[str] = None
    last_synced: Optional[datetime] = None
    event_count: int
    is_active: bool
    created_at: datetime

    model_config = {"from_attributes": True}


class CalendarEventResponse(BaseModel):
    id: int
    feed_id: int
    uid: str
    title: str
    description: Optional[str] = None
    start_date: datetime
    end_date: Optional[datetime] = None
    all_day: bool
    location: Optional[str] = None
    created_at: datetime

    model_config = {"from_attributes": True}


# ── ICS File Upload schemas ──────────────────────────────────


class ICSEventPreview(BaseModel):
    """A single parsed event from an uploaded .ics file."""
    index: int
    summary: str
    dtstart: datetime
    dtend: Optional[datetime] = None
    description: Optional[str] = None
    location: Optional[str] = None


class ICSParseResponse(BaseModel):
    """Response from parsing an uploaded .ics file (preview before import)."""
    events: list[ICSEventPreview]
    total: int


class ICSImportRequest(BaseModel):
    """Request to import selected events from a previously parsed .ics file."""
    selected_indices: Optional[list[int]] = None


class ICSImportResponse(BaseModel):
    """Result of importing ICS events as tasks."""
    created_count: int
    skipped_count: int
    errors: list[str]

File: frontend/src/api/calendarImport.ts
import { api } from './client';

export interface CalendarFeed {
  id: number;
  url: string;
  name: string | null;
  last_synced: string | null;
  event_count: number;
  is_active: boolean;
  created_at: string;
}

export interface CalendarEvent {
  id: number;
  feed_id: number;
  uid: string;
  title: string;
  description: string | null;
  start_date: string;
  end_date: string | null;
  all_day: boolean;
  location: string | null;
  created_at: string;
}

export const calendarImportApi = {
  async addFeed(url: string): Promise<CalendarFeed> {
    const { data } = await api.post<CalendarFeed>('/api/import/calendar', { url });
    return data;
  },

  async listFeeds(): Promise<CalendarFeed[]> {
    const { data } = await api.get<CalendarFeed[]>('/api/import/calendar');
    return data;
  },

  async syncFeed(feedId: number): Promise<CalendarFeed> {
    const { data } = await api.post<CalendarFeed>(`/api/import/calendar/${feedId}/sync`);
    return data;
  },

  async deleteFeed(feedId: number): Promise<void> {
    await api.delete(`/api/import/calendar/${feedId}`);
  },

  async listEvents(params?: { start?: string; end?: string; feed_id?: number }): Promise<CalendarEvent[]> {
    const { data } = await api.get<CalendarEvent[]>('/api/calendar-events', { params });
    return data;
  },
};
